examples/push-repo

```diff
@@ -0,0 +1,265 @@
+#!/usr/bin/env python
+
+import argparse
+import json
+import logging
+import kobo.rpmlib
+import os
+
+from chexus import Client, BucketItem, TableItem
+
+LOG = logging.getLogger("push-repo")
+
+
+def get_sigkey(rpmfile):
+    """
+    If the rpm is signed, return the sigkey.
+    rpmfile: path to an rpm
+    """
+    rpmhead = kobo.rpmlib.get_rpm_header(rpmfile)
+    rpmsig = kobo.rpmlib.get_keys_from_header(rpmhead)
+
+    # release, beta2, devel, old-release
+    valid_sigs = ["fd431d51", "f21541eb", "a5787476", "37017186"]
+
+    if rpmsig == None:
+        raise RuntimeError("Invalid RPM signature: %s is unsigned" % rpmfile)
+    if rpmsig.lower() not in valid_sigs:
+        raise RuntimeError(
+            "Invalid RPM signature: %s is signed with invalid key"
+            % (rpmfile, rpmsig.lower())
+        )
+    return rpmsig
+
+
+def cdn_path(file_name, file_path, checksum):
+    """
+    Get the cdn_path for the given file_info.
+    file_name: full name
+    file_path: full path to file, unless file is in cwd
+    checksum: the file's checksum
+    """
+    if file_name.endswith(".rpm"):
+        file_info = kobo.rpmlib.parse_nvra(file_name)
+        # the given information for creating the cdn_path
+        # so as an example for the rpm
+        # ipa-admintools-4.4.0-14.el7_3.1.1.noarch.rpm
+        cdn_path = os.path.join(
+            "/origin/rpms",
+            # ipa-admintools part
+            file_info["name"],
+            # 4.4.0 version part
+            file_info["version"],
+            # 14.el7_3.1.1
+            file_info["release"],
+            # fd431d51 - used by signing tools
+            get_sigkey(file_path),
+            # ipa-admintools-4.4.0-14.el7_3.1.1.noarch.rpm
+            # which seems to be the actual rpm that sits within
+            # the directory.
+            file_name,
+        )
+    else:
+        cdn_path = os.path.join(
+            "/origin/files", "sha256", checksum[:2], checksum, file_name
+        )
+    return cdn_path
+
+
+def repo_path(local_path, file_name, root_dir, dest_repo, is_iso):
```

Comment:
I don't quite get what this code is for.  Why does the code need to care whether it's publishing an ISO or yum repo and why does it have to do any recalculation of the paths based on that?

I gave this a try with a simple directory structure like:

```
rmcgover-test/
└── dir1
    ├── file1.txt
    └── file2.txt
```

...where I pass ``--local-repo $PWD/rmcgover-test/ --remote-repo /content/test/rmcgover-test-repo1``, and it ends up flattening the uploaded repo, discarding the structure I had locally:

```
Would publish the following item to the 'exodus' table;
	{
    "from_date": null,
    "metadata": {},
    "object_key": "853ff93762a06ddbf722c4ebe9ddd66d8f63ddaea97f521c3ecc20da7c976020",
    "web_uri": "/content/test/rmcgover-test-repo1/file1.txt"
}
```

Why is it flattened like that?  Whatever tool creates these repos (e.g. Pulp) would generally expect that the repo structure it created is kept as-is and not changed in this way.

There's a related comment under "main".